nbviewer/providers/__init__.py

```diff
@@ -35,17 +50,58 @@ def provider_uri_rewrites(providers):
     return _load_provider_feature('uri_rewrites', providers)
 
 
-def _load_provider_feature(feature, providers):
+def _load_provider_feature(feature, providers, **handler_names):
     """Load the named feature from an ordered list of dotted-notation modules
        which each implements the feature.
 
        The feature will be passed a list of feature implementations and must
        return that list, suitably modified.
     """
+    # `handler_names` is the same as the `handler_names` attribute of the NBViewer class
+
+    # Ex: provider = 'nbviewer.providers.url' 
+    # provider.rsplit(',', 1) = ['nbviewer.providers', 'url']
+    # provider_type = 'url'
+    provider_types = [provider.rsplit('.', 1)[-1] for provider in providers]
+
+    if 'github' in provider_types:
+        provider_types.append('github_blob')
+        provider_types.append('github_tree')
+        provider_types.remove('github')
```

Comment:
Hmm. This function used to be a handful of lines long and agnostic as to any specific providers it was loading. Is there a better place to put this logic specific to loading github providers?

Reply:
After reading the code here a couple times, I'm having trouble understanding what the new logic in the function is meant to support. Can you summarize the change?

Reply:
The first thing to note was that this was basically a hacky thing I did to ensure backward compatibility, not because I thought it was a good design per se. (Actually this is probably the portion of the code I wrote which I hate the most, but that's another story.)

Basically my crude understanding was that the `providers` command line argument only allows one to specify things at the level of individual "providers", which to me seemed like a unit of abstraction encompassing an entire "media source" with accompanying submodule containing multiple handlers and URI rewrites. Whereas what I was trying to write was at the sub-level of individual handlers for a specific provider.

The idea was to make the customization of individual handlers possible without affecting at all the previous user experience/the existing API, for those people not trying to do that.

So the point of this essentially is to make it possible for someone to say "I only want to run the GitHub provider", and have both of the default handlers used by default, but also if someone wants to only run the GitHub provider and customize only one of the two handlers, or both, then they can also do that (or hypothetically they could customize both handlers and then not run the GitHub provider).

Again though I hate this design and I only see it as being provisional, since my main goal was to basically include an "alpha" of new functionality in a way that doesn't at all modify existing functionality. But I think in the future another approach will be needed to support multiple levels of "granularity of customization".

Reply:
The **TL;DR** is that there is a one-to-many mapping from the "GitHub provider" to the relevant customizable handlers (contrast with the URL and localfile providers which each only have one customizable handler), so this is just a hack to implement that one-to-many mapping in a backwards compatible way, i.e. without forcing any user who's level of abstraction is only at "providers" to have to think about this one-to-many mapping if they don't care about handlers.